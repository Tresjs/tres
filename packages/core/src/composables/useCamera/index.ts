import type { TresContext } from '../useTresContextProvider'

import type { ComputedRef, Ref } from 'vue'
import { computed, ref, watchEffect } from 'vue'
import { isCamera, isPerspectiveCamera } from '../../utils/is'
import type { TresCamera } from '../../types'

/**
 * Interface for the return value of the useCamera composable
 */
export interface UseCameraReturn {

  activeCamera: ComputedRef<TresCamera>
  /**
   * The list of cameras
   */
  cameras: Ref<TresCamera[]>
  /**
   * Register a camera
   * @param camera - The camera to register
   * @param active - Whether to set the camera as active
   */
  registerCamera: (camera: TresCamera, active?: boolean) => void
  /**
   * Deregister a camera
   * @param camera - The camera to deregister
   */
  deregisterCamera: (camera: TresCamera) => void
  /**
   * Set the active camera
   * @param cameraOrUuid - The camera or its UUID to set as active
   */
  setActiveCamera: (cameraOrUuid: string | TresCamera) => void
}

/**
 * Interface for the parameters of the useCamera composable
 */
interface UseCameraParams {
  sizes: TresContext['sizes']
}

/**
 * Composable for managing cameras in a Three.js scene
 * @param params - The parameters for the composable
 * @param params.sizes - The sizes object containing window dimensions
 * @returns The camera management functions and state
 */
export const useCameraManager = ({ sizes }: UseCameraParams): UseCameraReturn => {
  const cameras = ref<TresCamera[]>([])
  const activeCamera = computed<TresCamera>(() => cameras.value[0]) // the first camera is used to make sure there is always one camera active

  /**
   * Set the active camera
   * @param cameraOrUuid - The camera or its UUID to set as active
   */
  const setActiveCamera = (cameraOrUuid: string | TresCamera) => {
    const camera = isCamera(cameraOrUuid)
      ? cameraOrUuid
      : cameras.value.find((camera: TresCamera) => camera.uuid === cameraOrUuid)

    if (!camera) { return }

    const otherCameras = cameras.value.filter(({ uuid }) => uuid !== camera.uuid)
    cameras.value = [camera, ...otherCameras]
  }

  /**
   * Register a camera
   * @param camera - The camera to register
   * @param active - Whether to set the camera as active
   */
  const registerCamera = (camera: TresCamera, active = false): void => {
    if (cameras.value.some(({ uuid }) => uuid === camera.uuid)) { return }
    cameras.value.push(camera)

    if (active) {
      setActiveCamera(camera.uuid)
    }
  }

  /**
   * Deregister a camera
   * @param camera - The camera to deregister
   */
  const deregisterCamera = (camera: TresCamera): void => {
    cameras.value = cameras.value.filter(({ uuid }) => uuid !== camera.uuid)
  }

  /**
   * Update camera aspect ratios when the window size changes
   */
  watchEffect(() => {
    if (sizes.aspectRatio.value) {
      cameras.value.forEach((camera: TresCamera) => {
        if (isPerspectiveCamera(camera)) {
          camera.aspect = sizes.aspectRatio.value
          camera.updateProjectionMatrix()
        }
      })
    }
  })

  return {
    activeCamera,
    cameras,
    registerCamera,
    deregisterCamera,
    setActiveCamera,
  }
}
