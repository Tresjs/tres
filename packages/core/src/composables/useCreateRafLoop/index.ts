import { createEventHook, useRafFn } from '@vueuse/core'
import { Clock } from 'three'

export interface RafLoopContext { delta: number, elapsed: number }

/**
 * @param cycleFn the function that is called before the after event hook is triggered and after the before event hook is triggered.
 */
export const useCreateRafLoop = (cycleFn: () => void) => {
  const clock = new Clock()

  const eventHooks = {
    before: createEventHook<RafLoopContext>(),
    after: createEventHook<RafLoopContext>(),
  }

  const { pause, resume, isActive } = useRafFn(() => {
    const getContextWithClock = (): RafLoopContext => ({
      delta: clock.getDelta(),
      elapsed: clock.elapsedTime,
    })

    eventHooks.before.trigger(getContextWithClock())
    cycleFn()
    eventHooks.after.trigger(getContextWithClock())
  }, {
    immediate: false,
  })

  const start = () => {
    clock.start()
    resume()
  }

  const stop = () => {
    clock.stop()
    pause()
  }

  return {
    start,
    stop,
    isActive,
    onBeforeLoop: eventHooks.before.on,
    onLoop: eventHooks.after.on,
  }
}
