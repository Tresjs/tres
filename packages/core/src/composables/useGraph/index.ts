import type { MaybeRef } from '@vueuse/core'
import { buildGraph } from '../../utils/graph'
import { computed, toValue } from 'vue'
import type { TresObject } from '../../types'

export const useGraph = (object: MaybeRef<TresObject | undefined>) => {
  return computed(() => {
    const obj = toValue(object)
    if (!obj) { return undefined }
    return buildGraph(obj)
  })
}
