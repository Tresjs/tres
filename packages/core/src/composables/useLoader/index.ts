import type { UseAsyncStateOptions, UseAsyncStateReturn } from '@vueuse/core'
import { useAsyncState } from '@vueuse/core'
import type { Loader, LoadingManager } from 'three'
import type { MaybeRef } from 'vue'
import { onUnmounted, reactive, toValue, watch } from 'vue'

import type { TresObject } from '../../types'
import { disposeObject3D } from '../../utils/'

export interface LoaderMethods {
  setDRACOLoader: (dracoLoader: any) => void
  setMeshoptDecoder: (meshoptDecoder: any) => void
  setKTX2Loader: (ktx2Loader: any) => void
}

export type TresLoader<T> = Loader & Partial<LoaderMethods> & {
  load: (
    url: string,
    onLoad: (result: T) => void,
    onProgress?: (event: ProgressEvent<EventTarget>) => void,
    onError?: (err: unknown) => void
  ) => void
  loadAsync: (url: string, onProgress?: (event: ProgressEvent) => void) => Promise<T>
}

export type LoaderProto<T> = new (manager?: LoadingManager) => TresLoader<T>

export interface TresLoaderOptions<T, Shallow extends boolean> {
  manager?: LoadingManager
  extensions?: (loader: TresLoader<T>) => void
  initialValue?: T
  asyncOptions?: UseAsyncStateOptions<Shallow, any | null>
}

/**
 * Return type for the useLoader composable
 * @template T - The type of the loaded asset (e.g., GLTF, Texture, etc.)
 * @template Shallow - Whether to use shallow reactivity for better performance
 * @extends {UseAsyncStateReturn<T, [string], Shallow>} - Extends VueUse's useAsyncState return type
 */
export type UseLoaderReturn<T, Shallow extends boolean> = UseAsyncStateReturn<T, [string], Shallow> & {
  /**
   * Loads a new asset from the given path
   * @param path - The URL or path to the asset to load
   */
  load: (path: string) => void
  /**
   * Progress of the loading process
   * @property loaded - The number of bytes loaded
   * @property total - The total number of bytes to load
   * @property percentage - The percentage of the loading process
   */
  progress: {
    loaded: number
    total: number
    percentage: number
  }
}

/**
 * Vue composable for loading 3D models using Three.js loaders
 * @param Loader - The Three.js loader constructor
 * @param path - The path to the model file
 * @param options - Optional configuration for the loader
 * @returns UseAsyncState composable with the loaded model
 */
export function useLoader<T, Shallow extends boolean = false>(
  Loader: LoaderProto<T>,
  path: MaybeRef<string>,
  options?: TresLoaderOptions<T, Shallow>,
): UseLoaderReturn<T, Shallow> {
  const proto = new Loader(options?.manager)
  const progress = reactive({
    loaded: 0,
    total: 0,
    percentage: 0,
  })
  if (options?.extensions) {
    options.extensions(proto)
  }

  const initialPath = toValue(path)
  const result = useAsyncState(
    (path?: string) => new Promise((resolve, reject) => {
      const assetPath = path || initialPath || ''

      proto.load(assetPath, (result: T) => {
        resolve(result as unknown as TresObject)
      }, (event: ProgressEvent<EventTarget>) => {
        progress.loaded = event.loaded
        progress.total = event.total
        progress.percentage = ((progress.loaded / progress.total) * 100)
      }, (err: unknown) => {
        reject(err)
      })
    }),
    options?.initialValue ?? null,
    {
      ...options?.asyncOptions,
      immediate: options?.asyncOptions?.immediate ?? true,
    },
  )

  // Watch for path changes and reload the model
  const unsub = watch(() => toValue(path), (newPath) => {
    if (newPath) {
      const value = result.state.value
      // Safely dispose the scene if it exists
      if (value && typeof value === 'object' && 'scene' in value && value.scene) {
        disposeObject3D(value.scene as unknown as TresObject)
      }
      result.execute(0, newPath)
    }
  })

  // Cleanup on component unmount
  onUnmounted(() => {
    unsub()
    const value = result.state.value
    if (value && typeof value === 'object' && 'scene' in value && value.scene) {
      disposeObject3D(value.scene as unknown as TresObject)
    }
  })

  return {
    ...result,
    load: (path: string) => {
      result.execute(0, path)
    },
    progress,
  } as UseLoaderReturn<T, Shallow>
}
