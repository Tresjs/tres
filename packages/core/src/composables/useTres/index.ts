import type { ComputedRef } from 'vue'
import type { TresContext } from '../useTresContextProvider'
import { useTresContext } from '../useTresContextProvider'
import type { Camera } from 'three'
import type { TresRenderer } from '..'

export interface TresPartialContext extends Omit<TresContext, 'renderer' | 'camera'> {
  /**
   * The renderer instance
   *
   * @type {TresRenderer}
   * @memberof TresPartialContext
   */
  renderer: TresRenderer
  /**
   * The current active camera
   *
   * @type {ComputedRef<Camera | undefined>}
   * @memberof TresPartialContext
   */
  camera: ComputedRef<Camera | undefined>
  /**
   * Marks the scene as needing an update in the next frame.
   * This is used in on-demand rendering mode to schedule a render.
   *
   * @type {() => void}
   * @memberof TresPartialContext
   */
  invalidate: () => void
  /**
   * Manually advances the render loop by one frame.
   * This is particularly useful in manual rendering mode where you want explicit control over when frames are rendered.
   *
   * @type {() => void}
   * @memberof TresPartialContext
   */
  advance: () => void
}

export function useTres(): TresPartialContext {
  const { scene, renderer, camera, sizes, controls, extend, events } = useTresContext()

  return {
    scene,
    renderer: renderer.instance,
    camera: camera.activeCamera,
    sizes,
    controls,
    extend,
    events,
    invalidate: renderer.invalidate,
    advance: renderer.advance,
  }
}
