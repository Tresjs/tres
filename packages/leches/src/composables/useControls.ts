import type { ToRefs } from 'vue'
import { isReactive, isRef, provide, reactive, ref, toRefs } from 'vue'
import type { Control } from '../types'

export const CONTROLS_CONTEXT_KEY = Symbol('CONTROLS_CONTEXT_KEY')
const DEFAULT_UUID = 'default'

// Internal state
const controlsStore: { [uuid: string]: { [key: string]: Control } } = reactive({})

export function useControlsProvider(uuid: string = DEFAULT_UUID) {
  provide(CONTROLS_CONTEXT_KEY, controlsStore)
  return controlsStore[uuid]
}
// Helper function to infer type
const inferType = (value: any): string => {
  const colorRegex = /^#(?:[A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$|^0x(?:[A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/

  if (typeof value === 'boolean') { return 'boolean' }
  if (typeof value === 'number') { return 'number' }
  if (typeof value === 'string' && colorRegex.test(value)) { return 'color' }
  if (typeof value === 'string') { return 'string' }
  if (value.isVector3
    || value.isVector2
    || value.isEuler
    || Array.isArray(value.value)
    || value.value.isVector3
    || value.value.isVector2
    || value.value.isEuler
    || Array.isArray(value.value.value)
  ) {
    return 'vector'
  }
  if (value.min !== undefined || value.max !== undefined || value.step !== undefined) { return 'range' }
  if (
    value.options
    && Array.isArray(value.options)) {
    return 'select'
  }

  // Add more types as needed
  return 'unknown'
}

const createControl = (key: string, value: any, type: string, folderName: string | null): Control => {
  const control: Control = {
    key: ref(key),
    label: ref(key),
    name: ref(key),
    type: ref(type),
    value: ref(value),
    visible: ref(true),
    icon: ref(),
    uniqueKey: ref(key),
    [key]: ref(value),
  }

  if (folderName) {
    control.folder = ref(folderName)
  }

  return control
}

export const dispose = (uuid: string = DEFAULT_UUID): void => {
  for (const key in controlsStore[uuid]) {
    delete controlsStore[uuid][key]
  }
}

export const useControls = (
  folderNameOrParams: string | { [key: string]: any },
  paramsOrOptions?: { [key: string]: any } | { uuid?: string },
  options?: { uuid?: string },
): Control | ToRefs<{ [key: string]: Control }> => {
  const result: { [key: string]: Control } = {}

  const folderName = typeof folderNameOrParams === 'string' ? folderNameOrParams : null
  const controlsParams = folderName ? paramsOrOptions as { [key: string]: any } : folderNameOrParams

  const actualOptions = folderName && folderName !== 'fpsgraph' ? options! : paramsOrOptions as { uuid?: string }
  const uuid = actualOptions?.uuid || DEFAULT_UUID

  if (!controlsStore[uuid]) {
    controlsStore[uuid] = reactive({})
  }

  if (folderNameOrParams === 'fpsgraph') {
    const control = createControl('fpsgraph', null, 'fpsgraph', null)
    controlsStore[uuid].fpsgraph = control
    result.fpsgraph = control
    return toRefs(reactive(result))
  }

  const controls = controlsStore[uuid]

  // Check if controlsParams is a reactive object
  const isParamsReactive = isReactive(controlsParams)
  const reactiveRefs = isParamsReactive ? toRefs(controlsParams as { [key: string]: any }) : {}

  for (const key in controlsParams as any) {
    let value = (controlsParams as any)[key]
    let uniqueKey = key

    // If controlsParams is reactive, use the reactive ref directly
    if (isParamsReactive && reactiveRefs[key]) {
      value = reactiveRefs[key]
    }

    // If the control is part of a folder, prefix the key with the folder name
    if (folderName) {
      uniqueKey = `${folderName}${key.charAt(0).toUpperCase() + key.slice(1)}`
    }

    // If the value is an object with control options
    if (typeof value === 'object' && !isRef(value) && !Array.isArray(value) && value.value !== undefined) {
      const controlOptions = value
      const reactiveValue = isRef(controlOptions.value) ? controlOptions.value : ref(controlOptions.value)
      const controlType = controlOptions.type || inferType(controlOptions)
      const control = createControl(key, reactiveValue, controlType, folderName)

      if (controlType === 'select') {
        control.options = ref(controlOptions.options.map((option: string | { text: string, value: any }) => {
          if (typeof option === 'object' && option.text && option.value) {
            return option
          }
          else {
            return {
              text: option,
              value: option,
            }
          }
        }))
      }

      if (controlType === 'range') {
        control.min = ref(controlOptions.min || 0)
        control.max = ref(controlOptions.max || 1)
        control.step = ref(controlOptions.step || 0.1)
      }

      control.label.value = controlOptions.label || key
      control.icon.value = controlOptions.icon
      control.visible.value = controlOptions.visible !== undefined ? controlOptions.visible : true
      control.uniqueKey.value = uniqueKey
      controls[uniqueKey] = control
      result[uniqueKey] = control
      continue
    }

    // If the value is a ref, use it directly
    if (isRef(value)) {
      const control = createControl(key, value, (value.value as any).type || inferType(value.value), folderName)
      controls[uniqueKey] = control
      result[uniqueKey] = control
      continue
    }

    // If the value is reactive, convert it to ref
    else if (typeof value === 'object' && !Array.isArray(value)) {
      const reactiveRefs = toRefs(value)
      if (reactiveRefs[key]) {
        value = reactiveRefs[key]
      }
    }

    // For non-ref values
    const control = createControl(key, value, value.type || inferType(value), folderName)

    // Update the internal state
    controls[uniqueKey] = control
    result[uniqueKey] = control
    control.uniqueKey.value = uniqueKey
  }

  return Object.keys(result).length > 1 ? toRefs(reactive(result)) : Object.values(result)[0]
}
