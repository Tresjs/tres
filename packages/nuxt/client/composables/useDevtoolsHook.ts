import type { TresObject, DevtoolsMessage } from '@tresjs/core'
import type { Scene } from 'three'
import type { SceneGraphObject } from '../types'
import { reactive, shallowReactive } from '#imports'
import { createSharedComposable } from '@vueuse/core'
import type { UnwrapNestedRefs } from 'vue'
import { getSceneGraph } from '../utils/graph'

// Define message data types for different message types
export interface ContextMessageData {
  scene: {
    value: Scene
  }
  renderer: {
    instance: {
      info: {
        render: {
          frame: number
          calls: number
          triangles: number
          points: number
          lines: number
        }
        memory: {
          geometries: number
          textures: number
        }
        programs: WebGLProgram[]
      }
    }
  }
}

export interface PerformanceMessageData {
  fps: FPSState
  memory: MemoryState
}

// Union type for all possible message types
export type DevtoolsMessageData = ContextMessageData | PerformanceMessageData

export interface FPSState {
  value: number
  accumulator: number[]
  lastLoggedTime: number
  logInterval: number
}

export interface MemoryState {
  currentMem: number
  averageMem: number
  maxMemory: number
  allocatedMem: number
  accumulator: number[]
  lastLoggedTime: number
  logInterval: number
}

export interface RendererType {
  fps: UnwrapNestedRefs<FPSState>
  memory: UnwrapNestedRefs<MemoryState>
  renderer: UnwrapNestedRefs<RendererState>
}

export interface RendererState {
  info: {
    render: {
      frame: number
      calls: number
      triangles: number
      points: number
      lines: number
    }
    memory: {
      geometries: number
      textures: number
    }
    programs: WebGLProgram[]
  }
}

export interface DevtoolsHookReturn {
  scene: {
    objects: number
    graph: SceneGraphObject | null
    value: Scene | undefined
    selected: TresObject | undefined
    /* assets: AssetInfo[] */
  }
  fps: FPSState
  memory: MemoryState
  renderer: RendererState
}

function countObjectsInScene(scene: Scene) {
  let count = 0

  scene.traverse((object) => {
    // Check if the object is a 3D object
    if (object.isObject3D) {
      count++
    }
  })

  return count
}

export interface DevtoolsState {
  scene: {
    objects: number
    graph: SceneGraphObject | null
    value: Scene | undefined
    selected: TresObject | undefined
    /* assets: AssetInfo[] */
  }
  fps: FPSState
  memory: MemoryState
  renderer: RendererState
}

function _useDevtoolsHook(): DevtoolsHookReturn {
  const state: DevtoolsState = {
    scene: shallowReactive({
      objects: 0,
      graph: null,
      value: undefined,
      selected: undefined,
      assets: [],
    }),
    fps: shallowReactive({
      value: 0,
      accumulator: [],
      lastLoggedTime: Date.now(),
      logInterval: 1000,
    }),
    memory: shallowReactive({
      currentMem: 0,
      averageMem: 0,
      maxMemory: 0,
      allocatedMem: 0,
      accumulator: [],
      lastLoggedTime: Date.now(),
      logInterval: 1000,
    }),
    renderer: reactive({
      info: {
        render: {
          frame: 0,
          calls: 0,
          triangles: 0,
          points: 0,
          lines: 0,
        },
        memory: {
          geometries: 0,
          textures: 0,
        },
        programs: [],
      },
    }),
  }

  let lastSceneUuid: string | null = null

  // Initialize DevtoolsMessenger if not exists
  if (typeof window !== 'undefined' && window.parent.parent.__TRES__DEVTOOLS__) {
    // Subscribe to different message types
    window.parent.parent.__TRES__DEVTOOLS__.subscribe((message: DevtoolsMessage<DevtoolsMessageData>) => {
      /* if (message.type === 'asset-load') {
        const messageData = message.data as AssetLoadData
        const formattedAsset = formatAsset(messageData)
        state.scene.assets.push(formattedAsset)
      } */

      if (message.type === 'context') {
        const context = message.data as ContextMessageData
        if (context.scene.value.children.length > 0) {
          // Use scene UUID for lightweight change detection
          const currentSceneUuid = context.scene.value.uuid
          if (currentSceneUuid !== lastSceneUuid) {
            state.scene.value = context.scene.value
            state.scene.objects = countObjectsInScene(context.scene.value)
            state.scene.graph = getSceneGraph(context.scene.value as unknown as TresObject)
            lastSceneUuid = currentSceneUuid
          }
        }
        else {
          state.scene.value = undefined
          state.scene.graph = null
          /* state.scene.assets = [] */
          lastSceneUuid = null
        }

        // Update renderer info from context
        if (context.renderer?.instance?.info) {
          Object.assign(state.renderer.info.render, context.renderer.instance.info.render)
          Object.assign(state.renderer.info.memory, context.renderer.instance.info.memory)
          state.renderer.info.programs = [...(context.renderer.instance.info.programs || [])]
        }
      }
      else if (message.type === 'performance') {
        const performance = message.data as PerformanceMessageData
        Object.assign(state.fps, performance.fps)
        state.fps.accumulator = [...performance.fps.accumulator]
        Object.assign(state.memory, performance.memory)
        state.memory.accumulator = [...performance.memory.accumulator]
      }
    })
  }

  return state
}

export const useDevtoolsHook = createSharedComposable(_useDevtoolsHook)
