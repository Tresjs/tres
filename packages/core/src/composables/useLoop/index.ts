import { useTresContext } from '..'
import { createPriorityEventHook } from '../../utils/createPriorityEventHook'
import type { RafLoopContext } from '../useCreateRafLoop'
import type { TresPartialContext } from '../useTres'
import { useTres } from '../useTres'

export type LoopContext = RafLoopContext & TresPartialContext

/**
 * Composable that provides control over the render loop and animation lifecycle.
 */
export const useLoop = () => {
  const tresContext = useTres()
  const { renderer: rendererManager } = useTresContext()

  const eventHookBeforeRender = createPriorityEventHook<LoopContext>()
  const eventHookAfterRender = createPriorityEventHook<LoopContext>()

  rendererManager.loop.onBeforeLoop((loopContext) => {
    eventHookBeforeRender.trigger({ ...tresContext, ...loopContext })
  })

  rendererManager.loop.onLoop((loopContext) => {
    eventHookAfterRender.trigger({ ...tresContext, ...loopContext })
  })

  const render = rendererManager.replaceRenderFunction

  return {
    stop: rendererManager.loop.stop,
    start: rendererManager.loop.start,
    isActive: rendererManager.loop.isActive,
    onBeforeRender: eventHookBeforeRender.on,
    onRender: eventHookAfterRender.on,
    render,
  }
}
