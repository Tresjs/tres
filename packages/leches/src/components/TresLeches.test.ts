import { mount } from '@vue/test-utils'
import { TresLeches } from '/@/'
import { expect, it } from 'vitest'
import { dispose } from '../composables/useControls'

afterEach(() => {
  dispose()
})

describe('tresLeches', async () => {
  it('should mount', async () => {
    const wrapper = mount(TresLeches, {
    /*    attachTo: document.body, */
      props: {
        uuid: 'test',
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('div').attributes('id')).toBe('test')
  })
})
