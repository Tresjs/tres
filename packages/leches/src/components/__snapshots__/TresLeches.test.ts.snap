// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TresLeches > should mount 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" id=\\"test\\" class=\\"tl-absolute tl-select-none tl-z-24 tl-w-280px tl-font-sans tl-text-xs\\">
  <div tabindex=\\"0\\" class=\\"tl-bg-white tl-shadow-xl tl-rounded tl-border-4 tl-border-solid tl-border-black\\">
    <header class=\\"tl-relative tl-cursor-grabbing tl-p-4 tl-flex tl-justify-between tl-text-gray-200 tl-relative\\"><i class=\\"tl-h-4 tl-w-4 tl-p-1.5 tl-flex tl-items-center tl-line-height-0 tl-rounded-full tl-bg-gray-100 tl-text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
  </div>
</div>"
`;

exports[`tresLeches > should mount 1`] = `
"<div style="left: 704px; top: 10px;" id="test" class="tl-absolute tl-select-none tl-z-24 tl-w-280px tl-font-sans tl-text-xs">
  <div tabindex="0" class="tl-bg-white tl-shadow-xl tl-rounded tl-border-4 tl-border-solid tl-border-black">
    <header class="tl-relative tl-cursor-grabbing tl-p-4 tl-flex tl-justify-between tl-text-gray-200 tl-relative"><i class="tl-h-4 tl-w-4 tl-p-1.5 tl-flex tl-items-center tl-line-height-0 tl-rounded-full tl-bg-gray-100 tl-text-xs">🍰</i>
      <div><i class="i-ic-baseline-drag-indicator"></i><i class="i-ic-baseline-drag-indicator"></i><i class="i-ic-baseline-drag-indicator"></i></div>
      <div></div>
    </header>
  </div>
</div>"
`;
